gcp/api/server.py

```diff
@@ -111,29 +111,45 @@ def QueryAffectedBatch(self, request, context):
   @ndb_context
   def DetermineVersion(self, request, context):
     """Determine the version of the provided hashes."""
-    if not request.query.name:
-      context.abort(grpc.StatusCode.NOT_IMPLEMENTED,
-                    'Querying only by file hash is not implemented yet.')
-      return None
-    return get_version_by_name(request.query).result()
+    return determine_version(request.query, context).result()
 
 
 @ndb.tasklet
-def get_version_by_name(
-    version_query: osv_service_v1_pb2.VersionQuery) -> ndb.Future:
-  """Identifies the version based on the provided name."""
-  query = osv.RepoIndex.query(osv.RepoIndex.name == version_query.name)
-  it = query.iter()
+def determine_version(version_query: osv_service_v1_pb2.VersionQuery,
+                      context: grpc.ServicerContext) -> ndb.Future:
+  """Identify fitting commits based on a subset of hashes"""
+  hashes = []
+  for i in range(min(20, len(version_query.file_hashes))):
+    hashes.append(version_query.file_hashes[i].hash)
+  idx_keys_set = set()
+  for h in hashes:
+    query = osv.RepoIndexResult.query(
+        osv.RepoIndexResult.file_results.hash == h)
+    query.keys_only = True
+    it = query.iter()
+    while (yield it.has_next_async()):
+      r = it.next()
+      idx_keys_set.add(r.key.parent())
+  if len(idx_keys_set) == 0:
+    context.abort(grpc.StatusCode.NOT_FOUND, 'no matches found')
+    return None
+
   futures = []
-  while (yield it.has_next_async()):
-    idx = it.next()
+  for r in idx_keys_set:
```

Comment:
Can we filter this down even more? i.e. to consider a `RepoIndex` at all, it needs to match **all** 20 hashes, rather than a single one. 

To do this, for the above, I think we can just track an intersection of candidates, rather than just `add`ing to a set.

We can also make this 20 number dynamic depending on the number of provided hashes.

Reply:
Yes, sorry this was only a first iteration of improvements.